File: app/src/main/java/io/ffem/lite/ui/ResultAdapter.java
package io.ffem.lite.ui;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;
import io.ffem.lite.R;
import io.ffem.lite.databinding.RowResultBinding;
import io.ffem.lite.model.TestResult;

import java.util.List;

public class ResultAdapter extends RecyclerView.Adapter<ResultAdapter.TestResultViewHolder> {

    @Nullable
    private List<? extends TestResult> mTestList;

    void setTestList(final List<? extends TestResult> testList) {
        if (mTestList != null) {
            mTestList.clear();
        }
        mTestList = testList;
        notifyItemRangeInserted(0, testList.size());
    }

    @NonNull
    @Override
    public TestResultViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RowResultBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.row_result,
                        parent, false);
        return new TestResultViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull TestResultViewHolder holder, int position) {
        holder.binding.setResult(mTestList.get(position));
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mTestList == null ? 0 : mTestList.size();
    }

    TestResult getItemAt(int i) {
        return mTestList.get(i);
    }

    static class TestResultViewHolder extends RecyclerView.ViewHolder {

        final RowResultBinding binding;

        TestResultViewHolder(RowResultBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/io/ffem/lite/ui/ResultAdapter.kt
package io.ffem.lite.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import io.ffem.lite.R
import io.ffem.lite.data.TestResult
import io.ffem.lite.databinding.RowResultBinding

class ResultAdapter(
    private val clickListener: (Int) -> Unit,
    private val longClickListener: (Int) -> Boolean
) :
    RecyclerView.Adapter<ResultAdapter.TestResultViewHolder>() {

    private var testList: List<TestResult>? = null

    internal fun setTestList(testList: List<TestResult>) {
        this.testList = testList
        notifyItemRangeInserted(0, testList.size)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TestResultViewHolder {
        val binding = DataBindingUtil
            .inflate<RowResultBinding>(
                LayoutInflater.from(parent.context), R.layout.row_result,
                parent, false
            )
        return TestResultViewHolder(binding)
    }

    override fun onBindViewHolder(holder: TestResultViewHolder, position: Int) {
        holder.binding.result = testList!![position]
        holder.binding.executePendingBindings()
        holder.binding.layout.setOnClickListener { clickListener(position) }
        holder.binding.layout.setOnLongClickListener { longClickListener(position) }
    }

    override fun getItemCount(): Int {
        return if (testList == null) 0 else testList!!.size
    }

    internal fun getItemAt(i: Int): TestResult {
        return testList!![i]
    }

    class TestResultViewHolder(val binding: RowResultBinding) :
        RecyclerView.ViewHolder(binding.root)
}
